Configurations/shared-info.pl

```diff
@@ -0,0 +1,102 @@
+#! /usr/bin/env perl
+# -*- mode: perl; -*-
+# Copyright 2016-2017 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the OpenSSL license (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+# This is a collection of extra attributes to be used as input for creating
+# shared libraries, currently on any Unix variant, including Unix like
+# environments on Windows.
+
+sub detect_gnu_ld {
+    my @lines =
+        `$config{cross_compile_prefix}$target{cc} -Wl,-V /dev/null 2>&1`;
+    return grep /^GNU ld/, @lines;
+}
+sub detect_gnu_cc {
+    my @lines =
+        `$config{cross_compile_prefix}$target{cc} -v 2>&1`;
+    return grep /gcc/, @lines;
+}
+
+my %shared_info;
+%shared_info = (
+    'gnu-shared' => {
+        shared_ldflag         => '-shared -Wl,-Bsymbolic',
+        shared_sonameflag     => '-Wl,-soname=',
+    },
+    'linux-shared' => sub {
+        return {
+            %{$shared_info{'gnu-shared'}},
+            shared_defflag    => '-Wl,--version-script=',
+        };
+    },
+    'bsd-gcc-shared' => sub { return $shared_info{'linux-shared'}; },
+    'bsd-shared' => sub {
+        return $shared_info{'gnu-shared'} if detect_gnu_ld();
+        return {
+            shared_ldflag     => '-shared -nostdlib',
+        };
+    },
+    'darwin-shared' => {
+        dso_lflags            => '-bundle',
+        shared_ldflag         => '-dynamiclib -current_version $(SHLIB_VERSION_NUMBER) -compatibility_version $(SHLIB_VERSION_NUMBER)',
+        shared_sonameflag     => '-install_name $(INSTALLTOP)/$(LIBDIR)/',
+    },
+    'cygwin-shared' => {
+        shared_ldflag         => '-shared -Wl,--enable-auto-image-base -Wl,-Bsymbolic',
+        shared_impflag        => '-Wl,--out-implib=',
```

Comment:
On side[!] note. I'm not actually fond of this kind of things. For following reason. There is a code somewhere that takes care of it. It ought to check if this variable is defined or not and act accordingly. It also ought to make an *assumption* that variable is defined in manner that permits appending a name. In other words complex logic nowhere near variable assignment. I would find it more appropriate if it was simply added to shared_ldflags with some kind of indirect reference to basename of currently processed library. Yes, it would mean that there would have to be code somewhere that processed the indirect reference, but that could be much more generic than checking for specific variable...

Reply:
Not sure how you figure the indirect reference to look like.  You can't use make variables, so that would be either to insert som perl code in there (I fear that would result in eval hell), or template stuff (perl within `{-` and `-}`, which leads to some kind of eval as well, but I actually find that less scary), or adding a sort of mini language (such as having `-Wl,--out-implib={LIB}` in `shared_ldflag`).
So while I value the idea (and "side[!] note" *is* noted), I think that's going to have to be for another PR, 'cause it needs some more thinking not to cause a larger mess.

Reply:
> that's going to have to be for another PR,

If *ever*! I mean I'm only saying I'm not *fond of* it and it might be worth considering something that would be more generalized :-)

Reply:
I understand that, and I actually agree...  thing is, something more generalized might end up more magic and with a load of complexity under the hood.  That's why I'm calling for caution.  But we should talk more about it (perhaps not right here, though, so we don't side track this PR too much)